UPGRADING_TO_OKHTTP_4.md

```diff
@@ -1,6 +1,74 @@
 Upgrading to OkHttp 4
 =====================
 
+OkHttp 4.x upgrades our implementation language from Java to Kotlin and keeps everything else the
+same. We've chosen Kotlin because it gives us powerful new capabilities while integrating closely
+with Java.
+
+We spent a lot of time and energy on retaining strict compatibility with OkHttp 3.x. We're even
+keeping the package name the same: `okhttp3`!
+
+There are three kinds of compatibility we're tracking:
+
+ * **Binary compatibility** is the ability to compile a program against OkHttp 3.x, and then to run
+   it against OkHttp 4.x. We're using use the excellent [japicmp][japicmp] library via its
+   [Gradle plugin][japicmp_gradle] to enforce binary compatibility.
+
+ * **Java source compatibility** is the ability to upgrade Java uses of OkHttp 3.x to 4.x without
+   making source code changes.
+
+ * **Kotlin source compatibility** is the ability to upgrade Kotlin uses of OkHttp 3.x to 4.x
+   without making source code changes.
+
+With one exception, OkHttp 4.x is both binary- and Java source-compatible with OkHttp 3.x. You can
+use an OkHttp 4.x .jar file with applications or libraries built for OkHttp 3.x. (The exception?
+`OkHttpClient` makes more things `final`.)
+
+OkHttp is **not** source-compatible for Kotlin callers, but upgrading should be automatic thanks to
+Kotlin's powerful deprecation features. Most developers should be able to use IntelliJ's _code
+cleanup_ for a safe and fast upgrade.
+
+For example, when we upgraded Square's Kotlin and Java codebases to OkHttp 4.x we had these
+problems:
+
+ * Single Abstract Method (SAM) conversions required us to replace lambdas with objects for Kotlin
+   callers.
+
+ * OkHttp 4.x's new `RequestBody.create()` overload conflicted with an overload in one of our
+   subclasses. The compiler caught this!
+
+ * The user-agent `okhttp/4.0.0-alpha02` was longer than the previous user-agent and a test failed
+   because the request's byte count changed.
```

Comment:
Is this worth mentioning? It won't be true of the final release.

Reply:
Intended to share how mundane the changes were. I'll drop it.